### PR TITLE
Remove blocks counting in copying rel pipeline

### PR DESCRIPTION
**Pull request by ray6080**

### CODE REVIEW

File: src/include/processor/operator/persistent/reader.h [hunk 1]
```diff
@@ -66,8 +65,7 @@ class Reader : public PhysicalOperator {
 
     read_rows_func_t readFunc;
     init_reader_data_func_t initFunc;
-    // For parallel reading.
-    std::unique_ptr<ReaderFunctionData> readFuncData;
+    std::shared_ptr<ReaderFunctionData> readFuncData;
```
Comment (acquamarin) [inline src/include/processor/operator/persistent/reader.h:68]: Can we store `readFuncData` in the globalState and keep raw pointers in the operator?

File: src/processor/operator/persistent/reader_state.cpp [hunk 2]
```diff
@@ -8,413 +8,115 @@ namespace kuzu {
...
         currBlockIdx = 0;
     }
-    return std::make_unique<ReaderMorsel>(currFileIdx, currBlockIdx++, currRowIdx);
+    if constexpr (READ_MODE == ReaderSharedState::ReadMode::PARALLEL) {
+        mtx.unlock();
+    }
+    return morsel;
 }
 
+template std::unique_ptr<ReaderMorsel>
```
Comment (acquamarin) [inline src/processor/operator/persistent/reader_state.cpp:116]: Should we move the declarations for template specialization to .h file?

**Review by acquamarin — APPROVED** [review]

File: src/processor/operator/persistent/reader_state.cpp (hunk 2 above)
Comment (ray6080) [reply src/processor/operator/persistent/reader_state.cpp:116]: This is not declaration, but instantiation.

File: src/include/processor/operator/persistent/reader.h (hunk 1 above)
Comment (ray6080) [reply src/include/processor/operator/persistent/reader.h:68]: We also need to keep a smart pointer for it when reader is parquet or npy reader.

**Pull request merged**